### PR TITLE
hmem, prov/efa: Add support for neuron dma-buf

### PR DESCRIPTION
**Pull request by aws-ceenugal**

### CODE REVIEW

File: prov/efa/src/efa_mr.c [hunk 1]
```diff
@@ -584,10 +584,63 @@ static struct ibv_mr *efa_mr_reg_ibv_mr(struct efa_mr *efa_mr, struct fi_mr_attr
...
+		}
+	} else {
+		goto ibv_reg_mr;
+	}
+
+	return ibv_mr;
+
+ibv_reg_mr:
+	return ibv_reg_mr(efa_mr->domain->ibv_pd,
+						(void *)mr_attr->mr_iov->iov_base,
```
Comment (shijin-aws) [inline prov/efa/src/efa_mr.c:637]: indent is off.

File: prov/efa/src/efa_mr.c [hunk 2]
```diff
@@ -584,10 +584,63 @@ static struct ibv_mr *efa_mr_reg_ibv_mr(struct efa_mr *efa_mr, struct fi_mr_attr
...
+						mr_attr->mr_iov->iov_len, access);
+}
 #else
 /**
  * @brief Register a memory buffer with rdma-core api.
- *
+ *
```
Comment (shijin-aws) [inline prov/efa/src/efa_mr.c:643]: unnecessary change

File: src/hmem_neuron.c [hunk 3]
```diff
@@ -89,6 +90,12 @@ static int neuron_dl_init(void)
 		goto err;
 	}
 
+	neuron_ops.nrt_get_dmabuf_fd = dlsym(neuron_handle, "nrt_get_dmabuf_fd");
+	if (!neuron_ops.nrt_get_dmabuf_fd) {
+        FI_WARN(&core_prov, FI_LOG_CORE, "Failed to find nrt_get_dmabuf_fd\n");
```
Comment (shijin-aws) [inline src/hmem_neuron.c:95]: indent.

**Comment by shijin-aws** [comment]
Thanks for the PR. It LGTM. Is neuron going to do p2p via dma-buf? Will that need a new efa-driver?

**Comment by shijin-aws** [comment]
bot:aws:retest

**Comment by shijin-aws** [comment]
Please look at the AWS CI failure, the code is not compiled
```
20:27:13 prov/efa/src/efa_hmem.c: In function 'efa_domain_hmem_info_init_neuron':
20:27:13 prov/efa/src/efa_hmem.c:241:60: error: 'dmabuf_fd' undeclared (first use in this function)
20:27:13   ret = neuron_get_dmabuf_fd((uint64_t)ptr, (uint64_t)len, &dmabuf_fd);
20:27:13                                                             ^~~~~~~~~
20:27:13 prov/efa/src/efa_hmem.c:241:60: note: each undeclared identifier is reported only once for each function it appears in
20:27:13 make[1]: *** [prov/efa/src/src_libfabric_la-efa_hmem.lo] Error 1
20:27:13 make[1]: *** Waiting for unfinished jobs....
20:27:13 make: *** [all] Error 2
20:27:13 
```

**Comment by aws-ceenugal** [comment]
@shijin-aws Thanks for reviewing. I'll address your comments soon

**Comment by shijin-aws** [comment]
Also the commit message: `prof`->`prov`

**Comment by shijin-aws** [comment]
Better to split this change into 2 commits. First commit for `src/hmem`, the second one for `prov/efa`.

File: prov/efa/src/efa_mr.c (hunk 1 above)
Comment (aws-ceenugal) [reply prov/efa/src/efa_mr.c]: Fixed

File: prov/efa/src/efa_mr.c (hunk 2 above)
Comment (aws-ceenugal) [reply prov/efa/src/efa_mr.c]: Removed

File: src/hmem_neuron.c (hunk 3 above)
Comment (aws-ceenugal) [reply src/hmem_neuron.c]: Fixed

**Comment by aws-ceenugal** [comment]
> Thanks for the PR. Some minor comments. Is neuron going to do p2p via dma-buf? Will that need a new efa-driver?

Yes, Neuron will perform p2p via dma-buf in the upcoming release (TBD). EFA-driver already supports dma-buf so we will not be needing any changes there.

**Comment by aws-ceenugal** [comment]
> Better to split this change into 2 commits. First commit for the common hmem code, the second one for `prov/efa`.

Done

**Comment by aws-ceenugal** [comment]
> Also the commit message: `prof`->`prov`

Done

**Comment by shijin-aws** [comment]
bot:aws:retest

**Review by shijin-aws — APPROVED** [review]

File: prov/efa/src/efa_mr.c [hunk 4]
```diff
@@ -584,6 +584,62 @@ static struct ibv_mr *efa_mr_reg_ibv_mr(struct efa_mr *efa_mr, struct fi_mr_attr
...
+		goto ibv_reg_mr;
+	}
+
+	return ibv_mr;
+
+ibv_reg_mr:
+	return ibv_reg_mr(
+		efa_mr->domain->ibv_pd,
+		(void *)mr_attr->mr_iov->iov_base,
+		mr_attr->mr_iov->iov_len, access);
```
Comment (wenduwan) [inline prov/efa/src/efa_mr.c:641]: May I suggest refactoring this block to avoid `goto`?

For example:
```c
if (!efa_mr_is_neuron(efa_mr))
    return ibv_reg_mr(...)

ret = neuron_get_dmabuf_fd(
				(uint64_t) mr_attr->mr_iov->iov_base,
				(uint64_t) mr_attr->mr_iov->iov_len,
				&dmabuf_fd);
if (ret != FI_SUCCESS) {
    // Actually I'm not sure if this should happen... why do we not error out instead of fallback to ibv_reg_mr?
}

ibv_mr = ibv_reg_dmabuf_mr(
						efa_mr->domain->ibv_pd, 0,
						mr_attr->mr_iov->iov_len,
						(uint64_t)mr_attr->mr_iov->iov_base,
						dmabuf_fd, access);
// Happy path

return ibv_mr;
```

Comment (aws-ceenugal) [reply prov/efa/src/efa_mr.c:641]: I will refactor the code. The fallback is to cover the case where Neuron driver doesn't support dmabuf feature (`neuron_get_dmabuf_fd` fails when the Neuron driver doesn't support dmabuf). This is for backward compatibility.

File: src/hmem_neuron.c [hunk 5]
```diff
@@ -89,6 +90,17 @@ static int neuron_dl_init(void)
 		goto err;
 	}
 
+	neuron_ops.nrt_get_dmabuf_fd = dlsym(neuron_handle, "nrt_get_dmabuf_fd");
```
[the remark this answers is not in the document]
Comment (aws-ceenugal) [reply src/hmem_neuron.c:93]: Correct, EFA 2.1.1.

File: prov/efa/src/efa_mr.c (hunk 4 above)
Comment (aws-ceenugal) [reply prov/efa/src/efa_mr.c]: Added a new commit with the refactored code. Please have a look.

**Review by wenduwan — APPROVED** [review]
Than you!

**Comment by shijin-aws** [comment]
bot:aws:retest

**Comment by shijin-aws** [comment]
Waiting for AWS CI.

**Comment by aws-ceenugal** [comment]
bot:aws:retest

**Comment by aws-ceenugal** [comment]
bot:aws:retest

**Comment by aws-ceenugal** [comment]
bot:aws:retest

**Comment by aws-ceenugal** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
@aws-ceenugal 

It looks like there are later commits that fix bug in eariler commits.

Can you re-organize your commits such that there are 2 commits: one for the hmem_neuron change, the other for `prov/efa` change?

The change to `hmem_neuron` should be titled "hmem_neuron: xxxx"

File: src/hmem_neuron.c [hunk 6]
```diff
@@ -178,6 +186,37 @@ void neuron_free(void **handle)
...
+
+	ret = neuron_ops.nrt_get_dmabuf_fd(va, size, fd);
+
+	if (ret == NRT_SUCCESS) {
+		return FI_SUCCESS;
+	} else if (ret == NRT_RESOURCE) { // real error from Neuron
+		FI_INFO(&core_prov, FI_LOG_CORE,
+			"Failed to retrieve dmabuf_fd: %d\n", ret);
+		return -FI_EINVAL;
+	} else { // fallback to mem registration using ibv_reg_mr
```
Comment (wenduwan) [inline src/hmem_neuron.c:213]: Comment style, the recommended practice is
```c
} else if (ret == NRT_RESOURCE) {
                /* Real error from neuron */
		FI_INFO(&core_prov, FI_LOG_CORE,
			"Failed to retrieve dmabuf_fd: %d\n", ret);
```

File: src/hmem_neuron.c [hunk 7]
```diff
@@ -178,6 +186,37 @@ void neuron_free(void **handle)
...
 
+/**
+ * @brief Get the dma-buf fd of Neuron memory region if it was registered for EFA peer direct
+ *
+ * @param addr[in] the device buffer address
+ * @param size[in] the device buffer size (in bytes)
+ * @param fd[out] the dma-buf fd
+ * @return int On success, return 0. On failure, return a negative error code
+ */
+int neuron_get_dmabuf_fd(uint64_t va, uint64_t size, int* fd) {
```
Comment (wenduwan) [inline src/hmem_neuron.c:197]: The new revision is easier to follow. Thank you!

File: prov/efa/src/efa_mr.c [hunk 8]
```diff
@@ -584,6 +584,48 @@ static struct ibv_mr *efa_mr_reg_ibv_mr(struct efa_mr *efa_mr, struct fi_mr_attr
...
+			mr_attr->mr_iov->iov_len, access);
+	}
+
+	int dmabuf_fd, ret;
+	ret = neuron_get_dmabuf_fd(
+			(uint64_t) mr_attr->mr_iov->iov_base,
+			(uint64_t) mr_attr->mr_iov->iov_len,
+			&dmabuf_fd);
+
+	if (ret == FI_SUCCESS) { // Success => invoke ibv_reg_dmabuf_mr
```
Comment (wenduwan) [inline prov/efa/src/efa_mr.c:611]: Same comment style comment.

**Review by wenduwan — APPROVED** [review]

**Comment by wenduwan** [comment]
bot:aws:retest

**Review by wzamazon — APPROVED** [review]

**Review by shijin-aws — APPROVED** [review]
